src/ht.c

```diff
@@ -1,239 +1,184 @@
- *    Keith Packard <keithp@keithp.com>
- * Integration in r2 core api and hackit up
- *    pancake <nopcode.org>
- */
-
 #include "ht.h"
 
-#define ARRAY_SIZE(array) (sizeof(array) / sizeof(array[0]))
-
-/*
- * From Knuth -- a good choice for hash/rehash values is p, p-2 where
- * p and p-2 are both prime.  These tables are sized to have an extra 10%
- * free to avoid exponential performance degradation as the hash table fills
- */
-
-static ut32 deleted_data;
-
-static const struct {
-	ut32 max_entries, size, rehash;
-} hash_sizes[] = {
-	{ 2,		5,		3	  },
-	{ 4,		7,		5	  },
-	{ 8,		13,		11	  },
-	{ 16,		19,		17	  },
-	{ 32,		43,		41        },
-	{ 64,		73,		71        },
-	{ 128,		151,		149       },
-	{ 256,		283,		281       },
-	{ 512,		571,		569       },
-	{ 1024,		1153,		1151      },
-	{ 2048,		2269,		2267      },
-	{ 4096,		4519,		4517      },
-	{ 8192,		9013,		9011      },
-	{ 16384,	18043,		18041     },
-	{ 32768,	36109,		36107     },
-	{ 65536,	72091,		72089     },
-	{ 131072,	144409,		144407    },
-	{ 262144,	288361,		288359    },
-	{ 524288,	576883,		576881    },
-	{ 1048576,	1153459,	1153457   },
-	{ 2097152,	2307163,	2307161   },
-	{ 4194304,	4613893,	4613891   },
-	{ 8388608,	9227641,	9227639   },
-	{ 16777216,	18455029,	18455027  },
-	{ 33554432,	36911011,	36911009  },
-	{ 67108864,	73819861,	73819859  },
-	{ 134217728,	147639589,	147639587 },
-	{ 268435456,	295279081,	295279079 },
-	{ 536870912,	590559793,	590559791 },
-	{ 1073741824,	1181116273,	1181116271},
-	{ 2147483648ul,	2362232233ul,	2362232231ul}
-};
-
-#define entry_is_free(x) (!x || !x->data)
-#define entry_is_deleted(x) (!x || x->data == &deleted_data)
-#define entry_is_present(x) (x && x->data && x->data != &deleted_data)
-
-/**
- * Finds a hash table entry with the given key and hash of that key.
- *
- * Returns NULL if no entry is found.  Note that the data pointer may be
- * modified by the user.
- */
-SdbHashEntry* ht_search(SdbHash *ht, ut32 hash) {
-	ut32 double_hash, hash_address;
-	if (ht && ht->entries) {
-		hash_address = hash % ht->size;
-		do {
-			SdbHashEntry *entry = ht->table + hash_address;
-			if (entry_is_free (entry))
-				return NULL;
-			if (entry_is_present (entry) && entry->hash == hash)
-				return entry;
-			double_hash = hash % ht->rehash;
-			if (double_hash == 0)
-				double_hash = 1;
-			hash_address = (hash_address + double_hash) % ht->size;
-		} while (hash_address != hash % ht->size);
+SdbHash* internal_ht_new(ut64 size, HashFunction hashfunction,
+		ListComparator comparator, DupKey keydup, DupValue valdup,
+		HtKvFreeFunc pair_free) {
+	SdbHash* ht;
+	ut32 i;
+	ht = calloc (1, sizeof(*ht));
+	if (ht != NULL) {
+		ht->size = size;
+		ht->count = 0;
+		ht->load_factor = 1;
+		ht->hashfn = hashfunction;
+		ht->cmp = comparator;
+		ht->dupkey = keydup;
+		ht->dupvalue = valdup;
+		ht->table = calloc (ht->size, sizeof (SdbList*));
+		if (ht->table != NULL) {
+			for (i = 0; i < ht->size; ++i) {
+				ht->table[i] = ls_new ();
```

Comment:
Cant we just keep those buckets to null until they need to be added to reduce memory and initial cpu usage

> On 23 Jun 2016, at 22:13, Jeffrey Crowell <notifications@github.com> wrote:
> 
> In src/ht.c:
> 
> > +		HtKvFreeFunc pair_free) {
> > +	SdbHash* ht;
> > +	ut32 i;
> > +	ht = calloc (1, sizeof(*ht));
> > +	if (ht != NULL) {
> > +		ht->size = size;
> > +		ht->count = 0;
> > +		ht->load_factor = 1;
> > +		ht->hashfn = hashfunction;
> > +		ht->cmp = comparator;
> > +		ht->dupkey = keydup;
> > +		ht->dupvalue = valdup;
> > +		ht->table = calloc (ht->size, sizeof (SdbList*));
> > +		if (ht->table != NULL) {
> > +			for (i = 0; i < ht->size; ++i) {
> > +				ht->table[i] = ls_new ();
> added to ls.h/c
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>